jax/_src/state/discharge.py

```diff
@@ -241,3 +251,473 @@ def _closed_call_discharge_rule(
                      else None for aval in in_avals)
   assert next(ref_vals_iter, None) is None
   return new_invals, out_vals
+
+# # `run_state`
+
+run_state_p = core.Primitive("run_state")
+run_state_p.multiple_results = True
+
+def _run_state_bind(*args: Any, jaxpr: core.Jaxpr,
+                    which_linear: tuple[bool, ...]):
+  if config.jax_enable_checks:
+    core.check_jaxpr(jaxpr)
+    assert len(jaxpr.invars) == len(args)
+    assert len(which_linear) == len(args)
+  return core.Primitive.bind(run_state_p, *args, jaxpr=jaxpr,
+                             which_linear=which_linear)
+run_state_p.def_custom_bind(_run_state_bind)
+
+def _run_state_impl(*args: Any, jaxpr: core.Jaxpr,
+                    which_linear: tuple[bool, ...]):
+  del which_linear
+  discharged_jaxpr, consts = discharge_state(jaxpr, ())
+  return core.eval_jaxpr(discharged_jaxpr, consts, *args)
+run_state_p.def_impl(_run_state_impl)
+
+def _run_state_abstract_eval(*avals: core.AbstractValue, jaxpr: core.Jaxpr,
+                             which_linear: tuple[bool, ...]):
+  del which_linear
+  # When we abstractly evaluate `run_state`, we want to keep track of which
+  # input avals are `Ref`s and which are not. If an aval is a `Ref`, we want to
+  # "propagate" out its inner effects. Otherwise, the effects are local to this
+  # `run_state`.
+  is_ref = set(i for i, aval in enumerate(avals)
+               if isinstance(aval, AbstractRef))
+  nonlocal_effects = set()
+  for eff in jaxpr.effects:
+    if isinstance(eff, RefEffect):
+      if eff.input_index not in is_ref:
+        continue
+    nonlocal_effects.add(eff)
```

Comment:
```suggestion
  nonlocal_effects = {e for e in jaxpr.effects if isinstance(e, RefEffect) and e.input_index in is_ref}
```

Reply:
I can't use this directly because we need to add *non*-RefEffects to `nonlocal_effects`.